# Week 1 - Dotplots and pairwise alignments

Data for practicals are located in html ~/Desktop/Bioinformatics. To update data run in terminal:

```bash
cd ~/Desktop/Bioinformatics
git pull
```
 
- [Sequence self-comparison](./sequence_alignment.md/#dotplots)
  - [Exercise 1.1](./sequence_alignment.md/#exercise-11---simple-self-comparison-using-dotplot)
  - [Exercise 1.2](./sequence_alignment.md/#exercise-12---identification-of-repetitive-motifs-using-dotplot)
- Compare two or more sequences
  - [Exercise 1.3](./sequence_alignment.md/#exercise-13---comparison-of-sequences-with-insertions-deletions-inversions)
  - [Exercise 1.8](./sequence_alignment.md/#exercise-18---identifying-overlaps-and-creating-a-sequence-assembly-using-dot-plots)
  - [Exercise 1.9](./sequence_alignment.md/#exercise-19---compare-two-genomic-regions-with-dotter)
- Whole genome comparison
  - [Exercise 1.10](./sequence_alignment.md/#exercise-110---whole-genome-comparison-with-gepard-program)
- Using dotter on protein sequences
  - [Exercise 1.4](./sequence_alignment.md/#exercise-14---comparison-of-her-proteins-using-dotplot)
  - [Exercise 1.5](./sequence_alignment.md/#exercise-15---comparison-of-hox-proteins-using-dotplot)
- Global vs. local alignment
  - [Exercise 2.2](./sequence_alignment.md/#exercise-22---pairwise-alignment-using-ncbi-blast)
