# Week 1 - Dotplots and pairwise alignments

Data for practicals are located in html ~/Desktop/Bioinformatics. To update data run in terminal:

```bash
cd ~/Desktop/Bioinformatics
git pull
```
 
- [Sequence self-comparison](./sequence_alignment.md/#dotplots)
  - [Exercise 1.1](./sequence_alignment.md/#exercise-11---simple-self-comparison-using-dotplot)
  - [Exercise 1.2](./sequence_alignment.md/#exercise-12---identification-of-repetitive-motifs-using-dotplot)
- Compare two or more sequences
  - [Exercise 1.3](./sequence_alignment.md/#exercise-13---comparison-of-sequences-with-insertions-deletions-inversions)
  - [Exercise 1.6](./sequence_alignment.md/#exercise-16---locate-exonintron-boundaries-using-dotter-splice-sites)
  - [Exercise 1.7](./sequence_alignment.md/#exercise-17---identification-of-insertions-deletions-duplications----more-complex-example)
  - [Exercise 1.8](./sequence_alignment.md/#exercise-18---identifying-overlaps-and-creating-a-sequence-assembly-using-dot-plots)
  - [Exercise 1.9](./sequence_alignment.md/#exercise-19---compare-two-genomic-regions-with-dotter)
  - [Exercise 1.11](./sequence_alignment.md/#exercise-111---identification-of-problems-in-sequences-from-sra-database)
- Whole genome comparison
  - [Exercise 1.10](./sequence_alignment.md/#exercise-110---whole-genome-comparison-with-gepard-program)
- Using dotter on protein sequences
  - [Exercise 1.4](./sequence_alignment.md/#exercise-14---comparison-of-her-proteins-using-dotplot)
  - [Exercise 1.5](./sequence_alignment.md/#exercise-15---comparison-of-hox-proteins-using-dotplot)
- Global vs. local alignment
  - [Exercise 2.1](./sequence_alignment.md/#exercise-21---compare-global-and-local-alignments)
  - [Exercise 2.2](./sequence_alignment.md/#exercise-22---pairwise-alignment-using-ncbi-blast)
  - [Exercise 2.3](./sequence_alignment.md/#exercise-23---comparing-p53-and-p63-protein-sequences)
- Multiple sequence alignment
  - [Exercise 3.1](./sequence_alignment.md/#exercise-31---multiple-sequence-alignment---cyclin-dependent-kinase)
  - [Exercise 3.2](./sequence_alignment.md/#exercise-32---multiple-alignment-from-hspb8-proteins)
  - [Exercise 3.3](./sequence_alignment.md/#exercise-33---alignment-of-protein-isoforms-alignment-editing)
  - [Exercise 3.4](./sequence_alignment.md/#exercise-34---multiple-sequence-alignment-of-the-mitochondrial-16s-gene)
  - [Exercise 3.5](./sequence_alignment.md/#exercise-35---msa-of-globins-proteins)
  - [Exercise 3.6](./sequence_alignment.md/#exercise-36---identification-of-catalytic-triad-residues-in-serine-proteases)
  - [Exercise 3.7](./sequence_alignment.md/#exercise-37---identification-of-bacterial-homologs-of-human-neuroglobin-and-analysis-of-heme-binding-pocket-conservation)
  - [Exercise 3.8](./sequence_alignment.md/#exercise-38---alignment-based-primer-design)