# Week 1 - Dotplots and pairwise alignments

Data for practicals are located in html ~/Desktop/Bioinformatics. To update data run in terminal:

```bash
cd ~/Desktop/Bioinformatics
git pull
```
## Dotplots
- [Dotplot demo I](https://kavonrtep.github.io/games/dotplot-demo.html)
- [Dotplot demo II](https://kavonrtep.github.io/games/dotplot-explorer.html)
- [FASTA format](./data_formats.md/#fasta-fa-fna-faa-fasta)
- [Sequence self-comparison](./sequence_alignment.md/#dotplots)
  - [Exercise 1.1](./sequence_alignment.md/#exercise-11---simple-self-comparison-using-dotplot)
  - [Exercise 1.2](./sequence_alignment.md/#exercise-12---identification-of-repetitive-motifs-using-dotplot)
  
- Whole genome comparison
  - [Exercise 1.10](./sequence_alignment.md/#exercise-110---whole-genome-comparison-with-gepard-program)

- Optional exercises
  - Using dotter on protein sequences
    - [Exercise 1.4](./sequence_alignment.md/#exercise-14---comparison-of-her-proteins-using-dotplot)
    - [Exercise 1.5](./sequence_alignment.md/#exercise-15---comparison-of-hox-proteins-using-dotplot)
  - Compare two or more sequences
    - [Exercise 1.8](./sequence_alignment.md/#exercise-18---identifying-overlaps-and-creating-a-sequence-assembly-using-dot-plots)
    - [Exercise 1.9](./sequence_alignment.md/#exercise-19---compare-two-genomic-regions-with-dotter)

# Week 2 
‚ùóBefore doing exercises run:

```bash
cd ~/Desktop/Bioinformatics
git pull
```
## Recap of dotplotb
 - [Excercise A1](./extra_tasks.md#exercise-a1)

## Dotplots - cont.
- Compare two or more sequences
  - [Exercise 1.3](./sequence_alignment.md/#exercise-13---comparison-of-sequences-with-insertions-deletions-inversions)

## Pairwise sequence alignment
- [Global alignment demo](https://kavonrtep.github.io/games/needleman-wunsch.html)
- [Local alignment demo](https://kavonrtep.github.io/games/smith-waterman.html)


## Multiple sequence alignment
- [Exercise 3.1 - Cyclin-dependent kinase](./sequence_alignment.md/#exercise-31---multiple-sequence-alignment---cyclin-dependent-kinase)
- [Exercise 3.3 - Alignment of protein isoforms, alignment editing](./sequence_alignment.md/#exercise-33---alignment-of-protein-isoforms-alignment-editing)


# Week 3

## Dotplot, local and global alignment
  - [Exercise 2.2](./sequence_alignment.md/#exercise-22---pairwise-alignment-using-ncbi-blast)
  - [Excercise A2](./extra_tasks.md#exercise-a2)

## Multiple sequence alignment - cont.
- [Exercise 3.2 - Multiple alignment from HSPB8 proteins](./sequence_alignment.md/#exercise-32---multiple-alignment-from-hspb8-proteins)
- [Exercise 3.6 - Identification of catalytic triad residues in serine proteases](./sequence_alignment.md/#exercise-36---identification-of-catalytic-triad-residues-in-serine-proteases)
- [Exercise 3.8 - Alignment-Based Primer Design](./sequence_alignment.md/#exercise-38---alignment-based-primer-design)
- [Exercise 3.5 - MSA of globins proteins](./sequence_alignment.md/#exercise-35---msa-of-globins-proteins)


# Week 4

## Multiple sequence alignment - cont.

- [Exercise 3.7 - Identification of Bacterial Homologs of Human Neuroglobin and Analysis of Heme-Binding Pocket Conservation](./sequence_alignment.md/#exercise-37---identification-of-bacterial-homologs-of-human-neuroglobin-and-analysis-of-heme-binding-pocket-conservation)

## NCBI BLAST

### Using BLAST from NCBI web interface
- [Exercise 1.1 - BLASTN - NCBI web interface](./blast_search.md/#exercise-11---blastn---ncbi-web-interface)
- [Exercise 1.2 - BLASTP - NCBI web interface](./blast_search.md/#exercise-12---blastp---ncbi-web-interface)
- [Exercise 1.3 - BLASTX - NCBI web interface](./blast_search.md/#exercise-13---blastx---ncbi-web-interface)
- [Exercise 1.6 - Identification of species using NCBI BLAST](./blast_search.md/#exercise-16---identification-of-species-using-ncbi-blast)
- [Exercise 1.4 - Identification of coding sequence using BLASTX](./blast_search.md/#exercise-14---identification-of-coding-sequence-using-blastx-vs-blastn)
- [Exercise 1.7 - Identification of mutations in gyrA gene](./blast_search.md/#exercise-17---identification-of-mutations-in-gyrA-gene)

### Using BLAST from command line (optional)
 - [Exercise 2.1](./blast_search.md/#exercise-21)
 - [Exercise 2.2](./blast_search.md/#exercise-22---extract-hits-from-database-and-create-alignment-with-query)

# Week 5 
## Genome assembly
- [Exercise 2 - Genome Assembly - Short Reads](./sequence_assembly.md#part-1---assembly-using-single-end-short-reads)
- [Exercise 2 - Genome Assembly - Short Paired-End Reads](./sequence_assembly.md#part-2---assembly-using-pair-end-short-reads)

# Week 6 
- [Exercise 2 - Genome Assembly - Long Reads](./sequence_assembly.md#part-3---assembly-using-long-pacbio-hifi-reads)

# Week 7 
- Genome Annotation

# Week 8
- Chip-Seq


# Week 9

- Variant calling
- Amplicon sequencing

# Week 10
- Creating phylogenetic tree
  - Exercise 1 - Comparison of phylogenetic trees based on amino acid and nucleotide sequences of ubiquitin gene
  - Exercise 2 -The case of the Florida dentist
  - Exercies 3 - Origin of HIV

# Week 11
- Structural bioinformatics, AlphaFold
